lib: os: onoff: Add cancellation before started feature

nordic-krch opened the pull request:
Extended onoff service to be capable of stopping service during starting.

Fixes #22974.

nordic-krch committed:
lib: os: onoff: Add cancellation before started feature

Extended onoff service to be capable of stopping service
during starting.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: onoff: Add test for forced cancelation

Added tests for cancelation during transition to on and off.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -142,6 +144,15 @@ static void onoff_start_notify(struct onoff_service *srv,
 	sys_slist_t clients = srv->clients;
 
+	/* Cancellation occurred, start notification is irrelevant */
+	if ((srv->flags & SERVICE_STATE_MASK) == SERVICE_STATE_TO_OFF) {
```
Protection against case when `onoff_cancel` was called and transition to off initiated but before cancel function is called transition to on occurs. In that case, it shall be skipped because service will be cancelled and notification from cancellation will come to the user.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -289,7 +301,7 @@ static void onoff_stop_notify(struct onoff_service *srv,
 	 * waiting clients of operation completion.
 	 */
-	if (res < 0) {
+	if ((res < 0) && (res != -ECANCELED)) {
```
`-ECANCELED` is not error from the service but from cancellation. In that case, it can transition to off state as if successful stop would occur. If new requests came in the meanwhile it will behave the same way as if they came during stopping.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -478,10 +490,16 @@ int onoff_service_reset(struct onoff_service *srv,
 
+static void onoff_cancel_notify(struct onoff_service *srv,  int res)
+{
+	onoff_stop_notify(srv, res >= 0 ? -ECANCELED : res);
```
at this point successful cancellation is like stop but with specific error code.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -494,14 +512,22 @@ int onoff_cancel(struct onoff_service *srv,
-			rv = -EWOULDBLOCK;
-			sys_slist_append(&srv->clients, &cli->node);
+			if (srv->cancel && (state == SERVICE_STATE_TO_ON)) {
+				set_service_state(srv, SERVICE_STATE_TO_OFF);
```
At this point, client is the last releaser, list is empty to accept new requests. transition to off is initiated. I'm not sure if `(state == SERVICE_STATE_TO_ON)`? If client is on the list, does it always mean that we are in volatile state (on to off or off to on)?

nordic-krch reviewed (commented):
Added few explanation comments.

pabigot reviewed (commented):
OK, thanks.  I think the problem is that you're not using the onoff manager the way it was intended.

https://github.com/zephyrproject-rtos/zephyr/pull/21090#issuecomment-571542437 in reply to https://github.com/zephyrproject-rtos/zephyr/pull/21090#issuecomment-571374676 notes that the manager is not designed as an end-user interface.  The expectation has always been that you'd have an intermediary, like a generic clock control API, that abstracts the details and provides extra functionality, including but not limited to support for user-mode.  This is true of all the manager APIs.

It's the lack of any intermediary service that pushes the need to handle special cases down to each individual client, which I agree is not acceptable.  I also agree that having something handle this case generically is important.

I have several concerns about changing the core onoff manager cancellation capability at this time.  I still believe most transitions can't just be cancelled: they will have to complete, then the reverse transition be started from a known state.  The solution proposed here is also unidirectional (doesn't support cancelling a stop, which is useful for any service that buffers data that has to be flushed before shutting down).  It would also be nice to support an arbitrary number of pending request/release operations with only the last desired transition tracked.

There is some potential for supporting a cancellation function that aborts an in-progress transition, but it needs to be bidirectional and the semantics and impact on registered notification objects much more tightly defined.  However, I think the use case of #22974 is best handled by another helper layer that hides the entire async state management from a user that needs to gate functionality on multiple conditions.  Such a helper is the `onoff_notifier` infrastructure proposed in #23229.  I'd like to see if that can be made to work for your situation.

nordic-krch commented:
> I still believe most transitions can't just be cancelled

yet, we have example of that in the first attempt of onoff usage.

> it needs to be bidirectional

I've added that. I've renamed those function to `uncond_start` and `uncond_stop` since the are not really cancel, it's rather capability of starting/stopping service disregarding its current state.

nordic-krch committed:
lib: os: onoff: Add support for canceling last item

Added support to cancel last request or release. Service is managing return
to expected state. It is also possible to cancel cancelation. For example,
if request is canceled and that is followed by next cancel then it will
result in system landing in on state and client notification with the result.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: onoff: Add tests for canceling scenarios

Extended suite to test various cases when cancelling might be called.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
